go/arrow/ipc/writer.go

```diff
@@ -587,16 +587,21 @@ func (w *recordEncoder) encodeMetadata(p *Payload, nrows int64) error {
 }
 
 func newTruncatedBitmap(mem memory.Allocator, offset, length int64, input *memory.Buffer) *memory.Buffer {
-	if input != nil {
-		input.Retain()
-		return input
+	if input == nil {
+		return nil
 	}
 
 	minLength := paddedLength(bitutil.BytesForBits(length), kArrowAlignment)
 	switch {
 	case offset != 0 || minLength < int64(input.Len()):
 		// with a sliced array / non-zero offset, we must copy the bitmap
-		panic("not implemented") // FIXME(sbinet): writer.cc:75
+		buf := memory.NewResizableBuffer(mem)
+		buf.Resize(int(minLength))
+		// only copy if some bits are set since CopyBitmap is slow
+		if bitutil.CountSetBits(input.Bytes(), int(offset), int(length)) > 0 {
+			bitutil.CopyBitmap(input.Bytes(), int(offset), int(length), buf.Bytes(), 0)
+		}
```

Comment:
I'd rather check `data.NullN() == data.Len()` above this function, it's cheaper and faster than `CountSetBits`